mutation_reader.cc

```diff
@@ -336,30 +358,59 @@ void mutation_reader_merger::retire_reader(reader_and_last_fragment_kind rk) {
-                // partitions we haven't even read yet.
-                // Readers whoose last emitted fragment was a partition
-                // end are out of data for good for the current range.
-                _halted_readers.push_back(rk);
+bool mutation_reader_merger::in_gallop_mode() const {
+    return _gallop_mode_hits >= gallop_mode_entering_threshold;
+}
+
+future<mutation_reader_merger::needs_merge> mutation_reader_merger::prepare_one(db::timeout_clock::time_point timeout,
+        reader_and_last_fragment_kind rk, handle_galloping handle_galloping) {
+    return (*rk.reader)(timeout).then([this, rk, handle_galloping] (mutation_fragment_opt mfo) {
+        if (mfo) {
+            if (mfo->is_partition_start()) {
+                _reader_heap.emplace_back(rk.reader, std::move(*mfo));
+                boost::push_heap(_reader_heap, reader_heap_compare(*_schema));
             } else {
-                retire_reader(rk);
+                if (handle_galloping) {
+                    // Optimization: assume that galloping reader will keep winning, and compare directly with the heap front.
+                    // If this assumption is correct, we do one key comparison instead of pushing to/popping from the heap.
+                    if (_fragment_heap.empty() || position_in_partition::less_compare(*_schema)(mfo->position(), _fragment_heap.front().fragment.position())) {
+                        _current.clear();
+                        _current.push_back(std::move(*mfo));
+                        _galloping_reader.last_kind = _current.back().mutation_fragment_kind();
+                        return needs_merge::no;
+                    }
+
+                    _gallop_mode_hits = 0;
+                }
+
+                _fragment_heap.emplace_back(rk.reader, std::move(*mfo));
+                boost::range::push_heap(_fragment_heap, fragment_heap_compare(*_schema));
             }
-        });
+        } else {
+            retire_reader(std::move(rk));
+            if (handle_galloping) {
+                // Galloping reader iterator may have become invalid at this point, so - to be safe - clear it
+                _galloping_reader.reader = { };
+            }
+        }
+
+        if (handle_galloping) {
+            _gallop_mode_hits = 0;
+        }
+        return needs_merge::yes;
+    });
+}
+
+future<mutation_reader_merger::needs_merge> mutation_reader_merger::advance_galloping_reader(db::timeout_clock::time_point timeout) {
+    return prepare_one(timeout, _galloping_reader, handle_galloping::yes).then([this] (needs_merge needs_merge) {
+        maybe_add_readers_at_partition_boundary();
+        return needs_merge;
+    });
+}
+
+future<> mutation_reader_merger::prepare_next(db::timeout_clock::time_point timeout) {
+    return parallel_for_each(_next, [this, timeout] (reader_and_last_fragment_kind rk) {
+        return prepare_one(timeout, rk, handle_galloping::no).discard_result();
```

Comment:
`discard_result()` attaches a continuation. Better just adjust the signature of the lambda below and ignore the parameter.

Reply:
`parallel_for_each` [accepts only futures which return nothing](https://github.com/scylladb/seastar/blob/master/include/seastar/core/future-util.hh#L194), so we are forced to call `.discard_result()` at the end.

Reply:
Right, missed that.